File: kubernetes/confluent/samples/nodejs/performance-80MB-throughput-producer.js
const { Kafka, ErrorCodes } = require('kafkajs');

// --- CONFIGURACIÓN ---
const TOPIC = 'test_throughput';
const MESSAGE_SIZE = 83886080; // 80MB
const DURATION = 60;           // 1 minuto en segundos

const kafka = new Kafka({
    clientId: 'throughput-tester',
    brokers: ['localhost:9092'], // Ajusta a tus brokers
    /* 
       Ajustes de reintentos y timeouts pesados para evitar que el cliente 
       desconecte si el broker tarda en procesar el bloque de 80MB
    */
    connectionTimeout: 10000,
    requestTimeout: 60000,
});

// --- MÉTRICAS ---
const latencies = [];

// Helper para calcular percentiles (reemplazo de numpy.percentile)
function getPercentile(arr, percentile) {
    if (arr.length === 0) return 0;
    const sorted = [...arr].sort((a, b) => a - b);
    const index = (percentile / 100) * (sorted.length - 1);
    const lower = Math.floor(index);
    const upper = Math.ceil(index);
    const weight = index - lower;
    if (upper >= sorted.length) return sorted[lower];
    return sorted[lower] * (1 - weight) + sorted[upper] * weight;
}

const formatNum = (num) => num.toLocaleString('en-US', { maximumFractionDigits: 2 });

async function runTest() {
    // Inicializar el productor
    const producer = kafka.producer({
        // Forzamos a que no intente agrupar lotes internos en JS duro
        maxInFlightRequests: 1
    });

    await producer.connect();

    console.log("Generando payload de 80MB en memoria...");
    const payload = Buffer.alloc(MESSAGE_SIZE, 'x');

    console.log(`--- Iniciando test (Duración: ${DURATION}s) ---`);
    console.log(`Enviando mensajes de ${formatNum(MESSAGE_SIZE / (1024 * 1024))} MB con kafkajs...`);

    const startTest = Date.now();
    const endTest = startTest + (DURATION * 1000);
    let aborted = false;

    // Capturar cierre controlado para desconectar el productor limpiamente
    const handleShutdown = async () => {
        if (aborted) return;
        aborted = true;
        console.log("\nPrueba finalizada. Desconectando productor...");
        await producer.disconnect();
        printReport(startTest);
        process.exit(0);
    };

    process.on('SIGINT', handleShutdown);
    process.on('SIGTERM', handleShutdown);

    try {
        // En kafkajs dependemos de un control asíncrono secuencial
        while (Date.now() < endTest && !aborted) {
            const startTime = Date.now();

            // Enviar el mensaje masivo de forma asíncrona pero secuencial
            await producer.send({
                topic: TOPIC,
                messages: [
                    { value: payload }
                ],
            });

            // Si llegó aquí, el broker confirmó la recepción (Equivalente al delivery_report)
            const latency = Date.now() - startTime;
            latencies.push(latency);
        }

        // Si el tiempo se cumple de forma natural
        await handleShutdown();

    } catch (err) {
        console.error(`\nError crítico durante el envío: ${err.message}`);
        await producer.disconnect();
    }
}

function printReport(startTest) {
    const totalTime = (Date.now() - startTest) / 1000;

    // --- CÁLCULO DE RESULTADOS ---
    if (latencies.length > 0) {
        const avgThroughput = latencies.length / totalTime;
        const avgLatency = latencies.reduce((a, b) => a + b, 0) / latencies.length;
        const p95Latency = getPercentile(latencies, 95);

        console.log("\n" + "=".repeat(40));
        console.log("         INFORME DE RENDIMIENTO (kafkajs)");
        console.log("=".repeat(40));
        console.log(`Mensajes confirmados: ${formatNum(latencies.length)}`);
        console.log(`Tiempo total:         ${totalTime.toFixed(2)} s`);
        console.log("-".repeat(40));
        console.log(`Throughput Medio:     ${avgThroughput.toFixed(2)} msg/s`);
        console.log(`Throughput MB/s:      ${((latencies.length * MESSAGE_SIZE) / (1024 * 1024) / totalTime).toFixed(2)} MB/s`);
        console.log("-".repeat(40));
        console.log(`Latencia Media:       ${avgLatency.toFixed(2)} ms`);
        console.log(`Latencia P95:         ${p95Latency.toFixed(2)} ms`);
        console.log("=".repeat(40));
    } else {
        console.log("\nNo se pudieron confirmar mensajes. Verifica los logs del broker.");
    }
}

// Ejecutar la prueba
runTest().catch(console.error);

File: kubernetes/confluent/samples/python/performance-80MB-throughput-producer.py
import time
import random
import string
import numpy as np
from confluent_kafka import Producer

# --- CONFIGURACIÓN ---
KAFKA_CONFIG = {
    'bootstrap.servers': 'localhost:9092', # Ajusta a tu broker
    'client.id': 'throughput-tester',
    # Control de buffer para evitar bloqueos de memoria
    'queue.buffering.max.messages': 10, 
    'linger.ms': 0,             # Agrupa mensajes para mejorar throughput
    'batch.num.messages': 1,  # Tamaño del lote
    'message.max.bytes': 90000000,
    # # 1. Configuración de Seguridad SSL/TLS
    # 'security.protocol': 'SASL_SSL',      
    # 'ssl.ca.location': '/ruta/a/ca-cert.pem', 
    
    # # 2. Configuración de Autenticación SASL 
    # 'sasl.mechanism': 'PLAIN',            
    # 'sasl.username': 'tu_usuario',
    # 'sasl.password': 'tu_contraseña',

}

TOPIC = 'test_throughput'
MESSAGE_SIZE = 83886080 # 80MB
DURATION = 60       # 1 minuto en segundos

# --- MÉTRICAS ---
latencies = []
total_msg_sent = 0

def delivery_report(err, msg, start_time):
    """Callback ejecutado tras la confirmación del Broker"""
    if err is not None:
        pass
    else:
        latency = (time.time() - start_time) * 1000
        latencies.append(latency)

def generate_payload(size):
    """Genera un string aleatorio de 10KB"""
    return ''.join(random.choices(string.ascii_letters + string.digits, k=size)).encode('utf-8')

def run_test():
    p = Producer(KAFKA_CONFIG)
    payload = generate_payload(MESSAGE_SIZE)
    
    print(f"--- Iniciando test (Duración: {DURATION}s) ---")
    print(f"Enviando mensajes de {MESSAGE_SIZE/1024}KB a: {KAFKA_CONFIG['bootstrap.servers']}")
    
    start_test = time.time()
    end_test = start_test + DURATION
    sent_count = 0

    try:
        while time.time() < end_test:
            current_time = time.time()
            try:
                # Intentar enviar mensaje
                p.produce(
                    TOPIC, 
                    payload, 
                    on_delivery=lambda err, msg, st=current_time: delivery_report(err, msg, st)
                )
                sent_count += 1
                
                # Servir eventos cada N mensajes para liberar el stack de callbacks
                if sent_count % 1000 == 0:
                    p.poll(0)

            except BufferError:
                # BACKPRESSURE: El buffer local está lleno. 
                # Esperamos a que Kafka procese lo pendiente antes de seguir.
                p.poll(0.1) 
                continue 

        print(f"\nTiempo cumplido. Procesando {p.flush()} mensajes restantes en cola...")
        p.flush() # Esperar confirmación de los últimos mensajes enviados

    except KeyboardInterrupt:
        print("\nPrueba cancelada por el usuario.")

    total_time = time.time() - start_test
    
    # --- CÁLCULO DE RESULTADOS ---
    if latencies:
        avg_throughput = len(latencies) / total_time
        avg_latency = np.mean(latencies)
        p95_latency = np.percentile(latencies, 95)
        
        print("\n" + "="*40)
        print("         INFORME DE RENDIMIENTO")
        print("="*40)
        print(f"Mensajes confirmados: {len(latencies):,}")
        print(f"Tiempo total:         {total_time:.2f} s")
        print("-" * 40)
        print(f"Throughput Medio:     {avg_throughput:.2f} msg/s")
        print(f"Throughput MB/s:      {(avg_throughput * MESSAGE_SIZE) / (1024*1024):.2f} MB/s")
        print("-" * 40)
        print(f"Latencia Media:       {avg_latency:.2f} ms")
        print(f"Latencia P95:         {p95_latency:.2f} ms")
        print("="*40)
    else:
        print("\nNo se recibieron confirmaciones del broker. Revisa la conexión.")

if __name__ == "__main__":
    run_test()